src/libraries/System.Runtime.Numerics/src/System/Numerics/BigNumber.cs

```diff
@@ -494,35 +494,243 @@ private static bool HexNumberToBigInteger(ref BigNumberBuffer number, out BigInt
+                            {
+                                char digitChar = fracDigitsSpan[i];
+                                if (digitChar == '\0')
+                                {
+                                    break;
+                                }
+                                if (digitChar != '0')
+                                {
+                                    return false;
+                                }
+                            }
+                        }
+
+                        partialValue = _partialValue;
+                        partialDigitCount = _partialDigitCount;
+                        totalDigitCount = _totalDigitCount;
+
+                        return true;
+                    }
+                }
+                else
                 {
-                    if (!ProcessChunk(digitsChunk.Span, ref currentBuffer))
+                    if (numberScale < 0)
                     {
                         result = default;
                         return false;
+
                     }
-                }
+                    totalDigitCount = Math.Min(number.digits.Length - 1, numberScale);
+                    int bufferSize = (totalDigitCount + MaxPartialDigits - 1) / MaxPartialDigits;
 
-                if (partialDigitCount > 0)
-                {
-                    MultiplyAdd(ref currentBuffer, s_uint32PowersOfTen[partialDigitCount], partialValue);
+                    Span<uint> buffer = new uint[bufferSize];
+
+                    int bufferInd = buffer.Length - 1;
+                    uint currentBlock = 0;
+                    int shiftUntil = (totalDigitCount - 1) % MaxPartialDigits;
+                    int remainingIntDigitCount = totalDigitCount;
+                    foreach (ReadOnlyMemory<char> digitsChunk in number.digits.GetChunks())
+                    {
+                        var digitsChunkSpan = digitsChunk.Span;
+                        ReadOnlySpan<char> intDigitsSpan = digitsChunkSpan.Slice(0, Math.Min(remainingIntDigitCount, digitsChunkSpan.Length));
+
+                        for (int i = 0; i < intDigitsSpan.Length; i++)
+                        {
+                            char digitChar = intDigitsSpan[i];
+                            Debug.Assert(char.IsDigit(digitChar));
+                            currentBlock = currentBlock * 10 + digitChar - '0';
+                            if (shiftUntil == 0)
+                            {
+                                buffer[bufferInd] = currentBlock;
+                                currentBlock = 0;
+                                bufferInd--;
+                                shiftUntil = MaxPartialDigits;
+                            }
+                            shiftUntil--;
+                        }
+                        remainingIntDigitCount -= intDigitsSpan.Length;
+
+                        ReadOnlySpan<char> fracDigitsSpan = digitsChunkSpan.Slice(intDigitsSpan.Length);
```

Comment:
I'm not particularly fond of the term `frac` (which I assume means `fractional`) here.

In the context of number parsing, it typically refers to the digits on the right hand side of the "decimal point", which should never exist for `BigInteger`.

Would `reaminingDigitsSpan` be a better name?

Reply:
Actually, its not quite clear to me why this is needed or why the other path has it either...

As indicated elsewhere, `BigInteger.Parse` treats number styles allowing exponents or decimal points as invalid, so this should never be possible.

Reply:
Here, I just rent the code from the [original one](https://github.com/dotnet/runtime/pull/55121/files#diff-83081f6bd1123c4813c96a452ead0a065d670bbf44ed85023f772253eaf7d2c5R567-R583), so I'm not sure about the variable name. I didn't feel any discomfort with this variable name, but if you feel any discomfort, I can fix it. In that case, I should fix the variable name in the original part at the same time.

Reply:
> Actually, its not quite clear to me why this is needed or why the other path has it either...

Given numbers are sometimes contain decimal points. I put the assert for the `fracDigitsSpan.Length == 0` here, then it failed.

Reply:
👍, this was just me not being awake enough when reviewing the code: https://github.com/dotnet/runtime/pull/55121#discussion_r715257253